### ISSUE TITLE
Support interdependencies between query parameters

### ISSUE THREAD

**Issue by cwarny** [OAI/OpenAPI-Specification#256]
It would be great to be able to specify interdependencies between query parameters. In my app, some query parameters become "required" only when some other query parameter is present. And when conditionally required parameters are missing when the conditions are met, the API fails. Of course I can have the API reply back that some required parameter is missing, but it would be great to have that built into Swagger.

**Comment by webron** [comment]
This is something that's very complicated to describe elegantly. Feel free to suggest a format that you think would be suitable for it.

**Comment by cwarny** [comment]
My suggestion would be as follows: instead of having the field "required" boolean, make it either boolean or an array of references to other parameters. If you put `required: true`, it would have to always be present. If you put `required: [$a, $b]`, it means that this parameter needs parameters `a` and `b` to be present.

**Comment by mohsen1** [comment]
How you describe an endpoint where one of it's parameters is required if another query parameter is equal to a string or is larger than a number? This can get very complicated. I would say if a parameter is required *sometimes* it's not required.

**Comment by cwarny** [comment]
I'd say either keep it simple and only allow simple interdependencies like, if `a` is present, then `b`is required (i.e. `required: [$a, $b]`), or define simple operations like, `required: [$a>3, $b=="hello", "world" in $c]`. I would favor keeping it simple, but then I have to admit that my only response to "why not keep it even simpler (i.e. as it is now) and only have a boolean `required`" is: it's a matter of how useful more sophistications are. For me, having simple interdependencies is useful and better than just a boolean `required`, but I don't think I need more complicated stuff. Other people might disagree.

**Comment by webron** [comment]
While this may be a solution, it can also be a validation nightmare. Definitely not something that can be done with a json schema.

**Comment by iisurge** [comment]
maybe something like
`"required": false,
"depends": [
$a,
$b
]`

**Comment by rcork** [comment]
Don't forget the "OR" dependency. I have a service where one of two query parameters must be present but it doesn't matter which one. For now we are just documenting the interdependency and using a error schema with the appropriate code/message but would be preferable to have this defined in the contract up front.

**Comment by fractalawareness** [comment]
+1 for **cwarny** simple suggestions

**Comment by filipnowakswissre** [comment]
My use-case is to have only one property of the parameter required, but from the array of the definitions.

**Comment by craig-bayley** [comment]
You can solve the OR use case already using an enumerated filter_type paired with a filter_value query parameter(s).

Example: 
        - name: filter_type
          in: query
          description: What type of values to filter on
          required: false
          type: string
          enum: [location, service, name]
          default: location
        - name: filter_value
          in: query
          description: The value to apply to the filter_type. Should match what was in the filter, i.e. if filtering on location use a location name.
          required: false
          type: string
          default: all

**Comment by filipnowakswissre** [comment]
wow, good one, thanks

**Comment by ePaul** [comment]
@craig-bayley could you format your example as YAML, please? Just adding

    ```yaml
    [...]
    ```
around it makes it much easier to read.

**Comment by craig-bayley** [comment]
@ePaul Thanks for the tip. Done.

**Comment by scottmas** [comment]
+1 Something like this would be very nice.

**Comment by fehguy** [comment]
Folks just to set expectations, we're not doing this in the initial cut of 3.0

**Comment by handrews** [comment]
> some query parameters become "required" only when some other query parameter is present

The JSON Schema `"dependencies"` keyword does exactly this.  The `"if"`/`"then"`/`"else"` keywords (json-schema-org/json-schema-spec#180) that seem likely to go into a near-future draft would also accomplish this (probably more clearly for most people).

**Comment by webron** [comment]
I don't see how JSON Schema would be relevant for this at all.

**Comment by handrews** [comment]
@webron further up you said:

> Definitely not something that can be done with a json schema.

so I assumed that meant that JSON Schema was at least _potentially_ relevant, which meant that I should point out that JSON Schema can in fact do this.  I think you were specifically referring to an approach based on `"required"` which would not be valid JSON Schema, but there is JSON Schema support for this use case.

**Comment by webron** [comment]
Gotcha. That referred to the validation of it, not the representation of it in the spec. I assumed you were referring to the representation, making it my bad.

**Comment by AlwaysAbhl001** [comment]
When it going to be on air ? much needed one.

**Comment by jainpradeep100** [comment]
I have need to generate @Context HttpHeader parameter by Swagger.  Is it possible or not? If possible then how it would be accomplish.

**Comment by webron** [comment]
@jainpradeep100 for tool-specific questions, please ask in its respective support channel, not here.

**Comment by HappyTreesDev** [comment]
Required could hold a simple true, false, or it could hold an object like:

```json
"required": [{
    "name" : "name of other param",
    "relationship" : "present, missing, or maybe some equation",
    "grouping" : "AND Vs OR"
}]
```

Something along the line of that where required would take an array of requirement objects that define the other relationships that the param has with other params.

**Comment by AndreKR** [comment]
I don't have enough experience with the OpenAPI specs to say if this is a good idea, but how about having an array of constraints, similar to how you would place CHECK constraints on a database schema.

Building upon [this](https://swagger.io/docs/specification/describing-parameters/#dependencies) example, something like:

```yaml
paths:
  /report:
    get:
      parameters:
        - name: rdate
          in: query
          schema:
            type: string
          description: >
             A relative date range for the report, such as `Today` or `LastWeek`.
             For an exact range, use `start_date` and `end_date` instead.
        - name: start_date
          in: query
          schema:
            type: string
            format: date
          description: >
            The start date for the report. Must be used together with `end_date`.
            This parameter is incompatible with `rdate`.
        - name: end_date
          in: query
          schema:
            type: string
            format: date
          description: >
            The end date for the report. Must be used together with `start_date`.
            This parameter is incompatible with `rdate`.
      constraints:
        - type: required_either_or
          either: [rdate]
          or: [start_date, end_date]
```

This way all kinds of relationships can be expressed, they can be extended in a backwards-compatible manner and shorthands for common scenarios make it intuitive to read.

**Comment by handrews** [comment]
I'm probably going to annoy people by being a broken record, but if `parameters` was a JSON Schema that was required to be of type "object", then you could express all of this in JSON Schema easily enough.  In the forthcoming draft-07, `"if"`/`"then"`/`"else"` makes it a lot more readable and easier to work with in code than implementing conditionals with `"oneOf"`.

**Comment by AndreKR** [comment]
I didn't notice I had picked an example where it's not a JSON Schema, in fact my `constraints` idea apply equally to for example `requestBody`, which is kind of compatibel with JSON Schema. So `parameters` not being a JSON Schema is a related but separate issue.

That said, my experience with JSON Schema and conditionals is limited to the abilities of [ajv-keywords](https://github.com/epoberezkin/ajv-keywords), but to be honest that was disappointing. I expected to be able to require and to forbid the existence of keys, values, default values and sub-schemas, using either the existence of keys or their values as a condition. It turned out half of those constraints are impossible to implement with ajv-keyword's if/then/else.

Is the JSON Schema proposal mightier?

**Comment by handrews** [comment]
@AndreKR ajv-keywords' if-then-else is the preliminary implementation, which got lots of positive feedback.  But please feel free to comment at https://github.com/json-schema-org/json-schema-spec/pull/375

I think we can work most of those out but let's not clutter up OAI's issue with that discussion :-)

**Comment by sivaramankannan** [comment]
This is something I have been looking for a long time. I like the idea of interoperating with the JSON schema for achieving this - please let me know the plans for this feature and if its on the cards.

**Comment by mewalig** [comment]
In the interest of flexibility... how about implementing the ability to define a third-party syntax (like jq)? From the perspective of the OpenAPI, the specification would merely require that it is specified as a third-party type, and contains the type identifier (e.g. "jq"), and corresponding filter/condition definition (e.g. the jq filter). Long shot but figured I'd ask...

**Comment by MikeRalphson** [comment]
@mewalig I appreciate you mentioned `jq` only as an example, but as far as I am aware, the `jq` syntax is not standardised anywhere, it only exists on the `jq` [website documentation pages](https://stedolan.github.io/jq/manual/), and `jq` itself is only available as a standalone program, not even a library for other tools to use.

That said, I'd be interested in seeing an example of the sort of syntax which you think might help with this issue. 

Nb: allowing multiple third-party syntaxes risks, as in #764, fragmenting the tooling community as to which tools support which syntaxes.

**Comment by mewalig** [comment]
@MikeRalphson I agree with those concerns and assume they stem from a requirement, were OpenAPI to go down this path, that the syntax of any supported third-party filter/condition/content be pre-defined and incorporated (directly or by reference) into the OpenAPI standard?

If so, one potential approach would be to offer this functionality as a further supported option of an [OpenAPI extension](https://swagger.io/docs/specification/openapi-extensions/) or some other way to essentially require various items to be provided in order for the third-party syntax to conform.

As a limited-scope example, a "jq" extension / package might require that the jq syntax is made available in some specified form. The term "made available" could be realized by any/all of
* included in the data itself. This would be most portable and flexible
* a URL to an OpenAPI-hosted location
* a URL to a non-OpenAPI-hosted location

Regardless of which of the above are supported, there would need to be a grammar specification that is defined in some form such as BNF or similar-- and there would need to be a "meta" grammar specification for the grammar specification. If the spec must be sufficient not just to validate rule syntax but also to validate the rule in the specific context in which it is being used, then the meta-grammar would also need to support a mechanism to identify items it must interpret, when validating, as data elements (for example, the jq rule ```.parameter.abc == "hello"``` is valid syntax, but would be invalid in the context of input that is not a dictionary with a member called "parameter" that is a dictionary type. If OpenAPI would require that this latter kind of error be detectable, the grammar spec would need some information to instruct the parser how to know that ".parameter.abc" refers to such a member).

Apart from specifying the grammar, applying a rule could be fairly straightforward. For example, a constraint requiring that the GET param "abc" equals "hello" might look something like the below:

```
paths:
  /report:
    get:
      parameters:
        abc:
          - type: array
          ...
      constraints:
        - type: extension
          name: arbitrary_name
          input_type: json # specify that the extension will consume the data as JSON. Could also be XML or other predetermined supported types
          content: .parameters | .abc[0] == "hello"
syntaxes:
  arbitrary_name:
    grammar: <grammar definition goes here>
```

**Comment by hkosova** [comment]
> @AndreKR commented on Aug 31, 2017:
> Building upon [this](https://swagger.io/docs/specification/describing-parameters/#dependencies) example, something like:
> ```yaml
> paths:
>   /report:
>     get:
>       parameters:
>         - name: rdate
>           in: query
>           schema:
>             type: string
>           description: >
>              A relative date range for the report, such as `Today` or `LastWeek`.
>              For an exact range, use `start_date` and `end_date` instead.
>         - name: start_date
>           in: query
>           schema:
>             type: string
>             format: date
>           description: >
>             The start date for the report. Must be used together with `end_date`.
>             This parameter is incompatible with `rdate`.
>         - name: end_date
>           in: query
>           schema:
>             type: string
>             format: date
>           description: >
>             The end date for the report. Must be used together with `start_date`.
>             This parameter is incompatible with `rdate`.
>       constraints:
>         - type: required_either_or
>           either: [rdate]
>           or: [start_date, end_date]
> ```

A possible workaround that exists in OpenAPI 3.0 is to define the interdependent parameters as a single object-type parameter with `style: deepObject`, and use the appropriate JSON Schema constraints (`oneOf`, `anyOf`, etc.) within the object schema.

```yaml
paths:
  /report:
    get:
      parameters:
        - in: query
          name: date_range
          required: true
          schema:   # either ?rdate=...  or ?start_date=...&end_date=...
            type: object
            oneOf:
              - properties:
                  rdate:
                    type: string
                    description: A relative date range, such as `Today`.
                required: [rdate]
                additionalProperties: false
              - properties:
                  start_date:
                    type: string
                    format: date
                  end_date:
                    type: string
                    format: date
                required: [start_date, end_date]
                additionalProperties: false
          style: deepObject
          explode: true
```
But this feels more of a hack than a real solution. Also, handling this in codegen will probably be tricky.

**Comment by handrews** [comment]
@hkosova 

> and use the appropriate JSON Schema constraints

This is exactly why URI Template parameters in JSON Hyper-Schema are handled with a single schema (the `hrefSchema` keyword plus some supporting modifiers).  It treats the URI Template as an object in the data model, and applies `hrefSchema` to the object as a whole.  So yes, you almost always end up having to say `{"properties": {...}"}` which is a tiny bit of annoying boilerplate, but it means that every single feature of JSON Schema works with parametrized URIs.  `hrefSchema` covers path and query parameters.

Likewise `headerSchema`, which would cover header and (indirectly) cookie parameters, although you can't mix it with `hrefSchema`.  So at least in the current draft you can't have dependencies between `hrefSchema` parameters and `headerSchema` parameters.

This would obviously have to be an OpenAPI 4 proposal, and even then might be too fundamentally against OAS's architectural principles, but the Parameter object could easily be replaced by a schema augmented with an extension vocabulary for for things like `deepObject` and other non-RFC 6570 / non-standard-JSON-Schema concepts.

**Comment by cernael** [comment]
Would allowing composition for the parameters object be an option? I'm thinking something like:
```yaml
paths:
  /pets/{pet_type}/{name}:
    get:
      summary: retrieve pet
      description: returns pet object
      parameters:
        oneOf:
          - 
            - name: pet_type
              in: path
              required: true
              description: type of pet
              schema:
                enum:
                  - cat
            - name: name
              in: path
              required: true
              description: name of cat
              schema:
                enum:
                  - Quaxo
                  - Corricopat
                  - Jellylorum
          - 
            - name: pet_type
              in: path
              required: true
              description: type of pet
              schema:
                enum:
                  - dog
            - name: name
              in: path
              required: true
              description: name of dog
              schema:
                enum:
                  - Buster
                  - Sture
                  - Old Yeller
```

**Comment by sleon-fmi** [comment]
This is a very common requirement in APIs I've encountered 'in the wild'. It would help a lot in describing legacy systems.

**Comment by mewalig** [comment]
I agree. However, I am not sure that the current approach can, or is even attempting to be able to, fully describe legacy APIs that did not have the OpenAPI limitations in mind when created. In my ideal world, I'd like for the API specification to be able to:
1. describe any number of validation expressions, each *of arbitrary complexity*
1. provide error messages corresponding to each of those validation expressions
1. support placeholders and substitutions within those error messages
1. support locale-specific error messages

Of these, to me, the first and second points are the most important, and it doesn't seem to me like it's being addressed in this project (though if OpenAPI supported a jq option or similar, I think it could be). That said, I'm not completely up-to-date on the OpenAPI developments so anyone, please correct me and/or clarify if I'm wrong or off base.

**Comment by handrews** [comment]
@mewalig number 1 is more of a JSON Schema thing than an OpenAPI thing.  As of OAS 3, parameter validation is done with OAS Schema Objects.

OAS 3.1 will have the `alternativeSchema` keyword, so if JSON Schema does not provide the sort of validation you want, it would be possible to find another system and register it as an available alternative.

Locale-specific strings in general may be best addressed by overlays (#1442, #1722).

It's a bit hard to figure out where 2 and 3 are best solved.  In part it depends on how you want to fill out the placeholders.  OAS is a design-time system so runtime resolution seems a bit out of scope.  If I'm understanding you correctly, anyway.

**Comment by kurt-o-sys** [comment]
I somewhat related, but slightly different case, modelling this in openapi:

```yaml
{
 "caseT": {... spec T...}
 "exts":
   {"caseU": {... spec U...},
    "caseX": {... spec X...}
  }
}
```

I have a number of (known) cases - this means, all possible keys are known (`case T`, `case U`, `case X`, `case ...`, this is limited). For each case, the spec is known as well (there's no problem with modelling the specs). The model here should be:

- there is exactly 1 of the cases in the 'root' of the request (in this example, this is case T),
- in an additional key `exts`, there are any number of other cases, but max 1 of each 

Going further: if possible at all any case except the one that's in the root (which I don't think is possible).
Or, I want to model a star schema: one central `case` and any number of other `case`s around that central `case`, e.g. (in my case, the central table/case is not known, it can be any of all known tables/cases):

![afbeelding](https://user-images.githubusercontent.com/2430465/53624141-81afeb80-3bff-11e9-87ce-a6524a77761f.png)

**Comment by mewalig** [comment]
Another potential approach would be to borrow Typescript syntax (which supports a broad range of types including [mutually-exclusive properties](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html)) to allow the creation and referencing of user-defined types that can be substituted for built-in types ("string", "number", "object" etc). This has the advantage of being a proven model for defining data types and would allow you to use a syntax that is already documented, together with a huge and growing opensource code base of typescript-related tools. In fact this would clearly provide numerous ways that OpenAPI could functionally scale. I'm having a hard time thinking of any drawbacks but I'm sure others can/will fill that gap...

**Comment by handrews** [comment]
@RexMagnus your use case is just:

```JSON
{
  "oneOf": [
    {
      "required": ["userId"],
      "not": {"required": ["email"]}
    },
    {
      "required": ["email"],
      "not": {"required": ["userId"]}
    }
  ]
}
```

**Comment by handrews** [comment]
@RexMagnus oh wait never mind, these are Parameter Objects :-(  
I lost track of which repository I was commenting on, sorry.  That only works inside of a Schema Object / JSON Schema.

**Comment by yami12376** [comment]
It's still open - after all these years?

**Comment by Somebi** [comment]
bump

**Comment by kryvel** [comment]
Any updates on this ?

**Comment by msudgh** [comment]
@webron 4 years open proposal! any updates?

**Comment by webron** [comment]
With no indication of importance, as far as I know, this is currently not on the radar of anyone from the @OAI/tsc. We have a list of topics we're more focused on for 3.1 described in #1466 

That said, we have a new proposals process any anyone is welcome to try making an official proposal so we can have a discussion on it. It's probably the best next step. There's no guarantee proposals would be accepted, but hopefully it can drive the discussion.

**Comment by Somebi** [comment]
bump

**Comment by darrelmiller** [comment]
Although there are some interesting suggestions in this thread, I don't see any clear proposal for addressing the raised concerns.  As @webron mentioned this is not an issue that we see raised commonly by the OAI community and as there are more pressing issues, I don't expect to see the @OAI/tsc doing any work on this in the near future.

**Comment by skretzig** [comment]
I've used the if/then/else pattern with JSON schema and it works like a charm. In my case, I'm trying to support versioning and backward compatibility in GET messages. If I can have conditional sets of parameters based on the value of the version query parameter, my problems are solved.
Just saying

**Comment by yigitsa** [comment]
TL;DR **_Wouldn't the simple case be resolved by allowing duplicate path mapping keys?_**

GET /resourceX/{somePathParam}
- email (required)
- username (required)


> There are many complex cases being discussed here, but my case is very straight forward and I think it's VERY common. 

I think @RexMagnus has a very valid point here, no need for complex cases. IMO if you need to write a DSL to define the interdependencies between query parameters, most likely you need to reconsider your design choices.

BUT,
Wouldn't this simple case be resolved by allowing duplicate path mapping keys?

For instance, if you can define 2 so-called `Path Item Objects` with the same HTTP verb (`Operation Object`) for the same `Paths Object`, then you can easily define parallel such "either both or none" query parameter groupings. 

And IMO it'd be much cleaner to understand from the client's POV - Otherwise the client needs to understand the logic of interdepende

**Comment by AML14** [comment]
So this is an open issue since January 2015. It has received 50 comments from more than 30 participants in the last 5 years, so there's clear interest in it. What's more, inter-parameter dependencies are _extremely_ common in RESTful APIs. That was our initial thought in my [research group](https://www.isa.us.es) and we have confirmed it by carrying out a thorough study on 40 real-world industrial APIs (~2600 operations), including those from YouTube, PayPal or Github itself, for instance.

We have recently published a paper regarding this issue at the [International Conference on Service-Oriented Computing](http://www.icsoc.org/) (ICSOC). A preprint is available [here](https://personal.us.es/amarlop/wp-content/uploads/2019/10/A-Catalogue-of-Inter-Parameter-Dependencies-in-RESTful-Web-APIs.pdf). We have also released the resulting dataset, which is available [here](https://bit.ly/2wvv1m1). We really hope these resources can provide a better understanding of the recurrence and dimension of this problem.

To summarize our findings, **85% of the APIs studied have inter-parameter dependencies**. They are present in all kinds of operations (GET, POST, etc.), even in those with very few parameters. Dependencies usually involve 2 query parameters, but we found uncommon cases like dependencies involving 10 parameters or involving parameters of different types (e.g. body and header parameters). All in all, we identified seven patterns of dependencies. All this is explained in the paper in detail.

**We are currently working on an OAS extension** for the expression of these seven patterns of dependencies, including, of course, all examples given by the participants in this thread. In our opinion, doing a revision of how dependencies emerge in practice was an essential step to tackle this problem. Now that we know what we are facing, we hope it'll be much easier to propose a solution.

**Comment by reuzel** [comment]
To keep things simple, why not have required support lists of parameter combinations of which one must be present? Something in the line of:
```
 required:
    - alwaysPresentParam
    - anyOf:
        - [ 'param1', 'param2' ]
        - [ 'param3' ]
 ```

**Comment by AML14** [comment]
> Why not have required support lists of parameter combinations of which oneOf/anyOf/allOf must be present? Required would implement allOf implicitly for backwards compatibility. If you can nest these, most of the combinations described in the article of @AML14 can be supported. It should be not too hard to create validation logic for this.
> 
> ```
>  required:
>     - alwaysPresentParam
>     - oneOf:
>         - mutuallyExclusiveParam1
>         - mutuallyExclusiveParam2
>     - oneOf:
>         - allOf:
>           - dependentParam1
>           - dependentParam2
>         - allOf:
>           - dependentParam3
>           - dependentParam4
>  
> ```

I think it's more complicated than that, @reuzel. What about the parameters that *forbid* the use of other parameters? Take for instance the `relatedToVideoId` parameter in the [search operation of the YouTube API](https://developers.google.com/youtube/v3/docs/search/list):

![image](https://user-images.githubusercontent.com/17610593/74030125-9455a780-49ae-11ea-93be-02d3284e219c.png)

If `relatedToVideoId` was combined with any parameter other than those eight listed there (e.g. `videoDefinition`), the API would return an error.

This is just one example, but there are many more use cases that need to be comtemplated. My team and I are working on it.

**Comment by MikeRalphson** [comment]
Glad to see this work and look forward to the proposals. I personally think it is unlikely OAS would adopt JSON Schema applicator keywords such as `allOf`, `oneOf`, `anyOf` or `not` (or properties with the same names) outside of schemaObjects, in order to avoid confusion.

**Comment by handrews** [comment]
I just proposed a correlation approach in #1635.  Here it is:

In OAS 3.1 we could solve this by adding a new OAS-specific schema extension keyword that could be used to correlate across parameters, headers, and bodies across requests and responses. You would use a `oneOf` in the schema of each thing that needs to correlate, possibly with an `if`/`then` in each branch if you want to explicitly call out the predicate for the server to test.

The extension keyword would be used in each `oneOf` branch to name it. Since that keyword would indicate that you are correlating things, an `if`/`then` alongside of the keyword would be taken as expressing the correlation predicate (there are other reasons you might use an if/then so it's useful to disambiguate the desired behavior).  You wouldn't necessarily need to use `if`/`then` in your `oneOf` branches, but some people find it more clear.

If you want to forbid certain parameters or headers in a particular branch, you can set their schemas to `false` (which is the same as `"not": {}` but with clearer intent- hmm... might not be able to support booleans in place of Schema Objects in 3.1 so might have to use `"not": {}`- either way it works).

That's all pretty hand-wavey. I can work up an example if there is interest.  But this basically lets you act like there's a big `oneOf` over the whole Operation Object, but without actually having to support JSON Schema applicator keywords outside of the Schema Object.

Of course we could also solve this by restructuring the Path Item or Operation Objects, but that would have to wait until OAS 4.

**Comment by ASHISHTOM** [comment]
> I just proposed a correlation approach in #1635. Here it is:
> 
> In OAS 3.1 we could solve this by adding a new OAS-specific schema extension keyword that could be used to correlate across parameters, headers, and bodies across requests and responses. You would use a `oneOf` in the schema of each thing that needs to correlate, possibly with an `if`/`then` in each branch if you want to explicitly call out the predicate for the server to test.
> 
> The extension keyword would be used in each `oneOf` branch to name it. Since that keyword would indicate that you are correlating things, an `if`/`then` alongside of the keyword would be taken as expressing the correlation predicate (there are other reasons you might use an if/then so it's useful to disambiguate the desired behavior). You wouldn't necessarily need to use `if`/`then` in your `oneOf` branches, but some people find it more clear.
> 
> If you want to forbid certain parameters or headers in a particular branch, you can set their schemas to `false` (which is the same as `"not": {}` but with clearer intent).
> 
> That's all pretty hand-wavey. I can work up an example if there is interest. But this basically lets you act like there's a big `oneOf` over the whole Operation Object, but without actually having to support JSON Schema applicator keywords outside of the Schema Object.
> 
> Of course we could also solve this by restructuring the Path Item or Operation Objects, but that would have to wait until OAS 4.



> I just proposed a correlation approach in #1635. Here it is:
> 
> In OAS 3.1 we could solve this by adding a new OAS-specific schema extension keyword that could be used to correlate across parameters, headers, and bodies across requests and responses. You would use a `oneOf` in the schema of each thing that needs to correlate, possibly with an `if`/`then` in each branch if you want to explicitly call out the predicate for the server to test.
> 
> The extension keyword would be used in each `oneOf` branch to name it. Since that keyword would indicate that you are correlating things, an `if`/`then` alongside of the keyword would be taken as expressing the correlation predicate (there are other reasons you might use an if/then so it's useful to disambiguate the desired behavior). You wouldn't necessarily need to use `if`/`then` in your `oneOf` branches, but some people find it more clear.
> 
> If you want to forbid certain parameters or headers in a particular branch, you can set their schemas to `false` (which is the same as `"not": {}` but with clearer intent).
> 
> That's all pretty hand-wavey. I can work up an example if there is interest. But this basically lets you act like there's a big `oneOf` over the whole Operation Object, but without actually having to support JSON Schema applicator keywords outside of the Schema Object.
> 
> Of course we could also solve this by restructuring the Path Item or Operation Objects, but that would have to wait until OAS 4.

when will OAS 4 will release??

**Comment by grexe** [comment]
back-linking relevant SO thread with some proposed workarounds for OAS 3.0:
https://stackoverflow.com/a/49199240/160799

**Comment by thealjey** [comment]
The only clean solution imo is to not have `parameters` defined as an array type, but instead make it work the same as `requestBody` does, then one would be free to use `oneOf` or anything they like.
On top of that, parsing and validating of such a shape would be easier, since you can just reuse the existing `requiestBody` logic.

**Comment by AML14** [comment]
Hi everyone. I'm happy to let you know that my team and I have finally made public our proposal for handling inter-parameter dependencies in OAS. It is a domain-specific language (DSL) called IDL (Inter-parameter Dependency Language), which supports the seven patterns of dependencies that I referred to in [my previous comment in this thread](https://github.com/OAI/OpenAPI-Specification/issues/256#issuecomment-547569202). We have also implemented a Java library for performing a number of analysis operations in these dependencies, e.g., to check if a given API request satisfies all dependencies.

It is all summed up in these two Medium posts:
1. [Inter-parameter dependencies in REST APIs](https://medium.com/isa-group/inter-parameter-dependencies-in-rest-apis-4664e901c124) (5-min read).
2. [Handling inter-parameter dependencies in REST APIs with IDL4OAS](https://medium.com/isa-group/handling-inter-parameter-dependencies-in-rest-apis-with-idl4oas-e5489abac32f) (7-min read).

While the second post describes the DSL and the analysis tool, the first one can also be useful for alternative implementations of both resources, since it describes the most common types of dependencies appearing in real-world APIs.

All this work is based on a paper recently accepted in the journal IEEE Transactions on Services Computing (preprint available [here](researchgate.net/publication/348280988_Specification_and_Automated_Analysis_of_Inter-Parameter_Dependencies_in_Web_APIs)).

Hope you enjoy it!

**Comment by MikeRalphson** [comment]
@AML14 congratulations on publishing your research findings and your proposal. Is there a BNF grammar for the DSL available? One thing which occurred to me is: can IDL distinguish between two parameters with identical names which only differ by their `in:` field - i.e. two parameters called `clientId` one `in: query` and one `in: header`, *one of which* must be present?

**Comment by AML14** [comment]
Thanks @MikeRalphson. You can have a look at the IDL grammar [here](https://github.com/isa-group/IDL/blob/master/es.us.isa.interparamdep/src/es/us/isa/interparamdep/InterparameterDependenciesLanguage.xtext). We also provide a simplified version of the grammar in the paper, attached at the end of this comment.

Regarding your question, it's a good point. Currently no, but just because it's not a common scenario. Since IDL is implemented with Xtext, it can be easily extended. For example, it could be expressed as follows:

```
OnlyOne([q]clientId, [h]clientId);
```

`[q]` and `[h]` would be optional elements, used when this information needs to be specified. The IDL grammar can be updated to support this syntax, and the IDL parser and code generator could thereby handle such a scenario.

<img src="https://user-images.githubusercontent.com/17610593/104244288-62707f00-5462-11eb-86ff-981c0e17ea61.png" width="50%" />

**Comment by AML14** [comment]
I just opened an issue in the OpenAPI Generator repository: https://github.com/OpenAPITools/openapi-generator

It is a feature request, which we are willing to start developing, but would like to get as much feedback from the community as possible. The point is: **Based on the IDL dependencies included in OAS, we could automatically generate built-in assertions in the source code of clients and servers (any programming language) that handle the validation of such dependencies**.

I think this was one of the main ideas when this issue was opened, but it's going to be big, so we need to discuss it all together and provide our own views on the matter.

**Comment by SteveNay-vz** [comment]
This issue is now over 6 years old. :-(  OAS 3.1 was just released yesterday. I don't want to do any disservice to the IDL solution offered by @AML14, but was _anything_ done to support inter-parameter dependencies in v3.1? 

I'm not overly familiar with JSON schema (now fully supported in 3.1), but I spent a little time reading about [dependencies](https://json-schema.org/understanding-json-schema/reference/object.html#dependencies) and "[applying subschemas conditionally](https://json-schema.org/understanding-json-schema/reference/conditionals.html)" (if/then/else), and I don't see how they can help with inter-parameter dependencies. Unless I missed something, both of those concepts only apply _within_ an object, but query parameters are completely separate objects, each with their own schema. 

If JSON schema can be used to describe relationships between query parameters, I would really appreciate it if someone can point me to where that is described and perhaps an example. Or, if IDL4OAS is the officially recommended method of describing parameter relationships, let's just say so and close this issue and then all go have a drink to celebrate.

**Comment by MikeRalphson** [comment]
@SteveNay-vz 

> I don't want to do any disservice to the IDL solution offered by @AML14, but was anything done to support inter-parameter dependencies in v3.1?

No, it wasn't a focus during development of v3.1. That development started in October 2018, I don't think any non-breaking proposals were made in that time-frame, except adding a specification-extension, which of course, anyone is free to do at any time. We're always interested in people's experiences with feature pilots. 

> I'm not overly familiar with JSON schema (now fully supported in 3.1), but I spent a little time reading about dependencies and "applying subschemas conditionally" (if/then/else), and I don't see how they can help with inter-parameter dependencies. Unless I missed something, both of those concepts only apply within an object, but query parameters are completely separate objects, each with their own schema.

You haven't missed anything. Several people above have suggested using a single JSON Schema object to represent query parameters, path parameters and headers. That poses some problems (do you have three objects or one, how are `style` and `explode` handled - as a new JSON Schema vocabulary?) but I suspect it will certainly be discussed for v4.

> Or, if IDL4OAS is the officially recommended method of describing parameter relationships, let's just say so and close this issue and then all go have a drink to celebrate.

IDL4OAS was only proposed during the last month of v3.1 development, when we were already into the RC phase. We haven't yet seen enough real-world feedback on it to even evaluate it properly, let alone recommend it. Darn, no drink for me. 😿 

> This issue is now over 6 years old. :-( OAS 3.1 was just released yesterday. 

Unless GitHub have removed helpful features, this is a little redundant.

**Comment by AML14** [comment]
IDL4OAS is only as powerful as we, as a community, want it to be(come). Thanks to it, we've been able to develop tools that leverage the inter-parameter dependencies in the API, for example:

1. [**IDLReasoner**](https://github.com/isa-group/IDLReasoner): a tool for performing multiple analysis operations such as checking if the dependencies are satisfied in a request. So, for instance, it could be deployed on an API gateway to reject all requests that violate dependencies, without redirecting those to the respective service.
2. [**RESTest**](https://github.com/isa-group/RESTest): a testing framework for REST APIs that generates requests satisfying/violating the inter-parameter dependencies of the web API operation. Internally, it uses IDLReasoner.

I just announced the opening of that issue in openapi-generator because I think automatic code generation is one of the greatest strengths of OpenAPI. If we can automatically generate the code in charge of checking if inter-parameter dependencies are satisfied or not in API clients and server stubs, that will be awesome.

**Comment by AML14** [comment]
For those interested in automatically generating built-in assertions in the source code of clients and servers, based on the IDL dependencies present in the OAS document (using our IDL4OAS extension), we just implemented a preliminary feature in OpenAPI Generator: https://github.com/OpenAPITools/openapi-generator/issues/8722.

It currently targets Java (several frameworks of clients and servers), but it could be easily extended to other languages and frameworks in the future, especially taking @enriquebarba97's [implementation](https://github.com/enriquebarba97/openapi-generator/tree/idl-dependencies) as a reference.

All feedback welcome. Hope this is the beginning of some new cool stuff for the OAS ecosystem 🥳.

**Comment by vellala2000** [comment]
issues like these are critical for Open API adoption. Hope we will have solution for this in next version of openapi release.

**Comment by jon-jon-jon** [comment]
Find my solution here https://github.com/OAI/OpenAPI-Specification/issues/182#issuecomment-961767168

**Comment by kaiserbergin** [comment]
Would be really rad to have some movement on this as it's fairly common to have interdependencies among parameters. Translating and interpreting what's required in those scenarios is a bit difficult in the current state, outside of comments in descriptions.

**Comment by karenetheridge** [comment]
This can be done in the JSON Schema, if all the query parameters are treated as one "parameter" and passed to the schema evaluator as an object type.

The parameter deserialization process is not specified very clearly, but I do something similar when handling header parameters -- if the "type" keyword is present at the top level of the json schema for the header, and it is "array", then multiple values of that header are permitted and they are passed to the evaluator as an array.  If it's "string", then only one header value is allowed; same thing for "number" or "integer", but in that case the parsed header value is also converted to a number (if possible) before being sent to the evaluator.

**Comment by alexyans** [comment]
Another kind of interdependency I would like to see in a future version is enumerations for combinations of parameters. Imagine for example a set of endpoints that are scoped to geographic locales, but some of the functionality or some of the resources are only available to certain locales.

It would be nice to be able to specify what combinations of parameters are acceptable, rather than just the values that each parameter can accept individually.

Then an API Gateway configured with Swagger can correctly intercept invalid requests without the need for validation at service level. In large distributed architectures it could eliminate a lot of repetitive validation logic.

**Comment by handrews** [comment]
This will be addressed by the OAS 4 Moonwalk project.  Active discussions include:

* https://github.com/OAI/moonwalk/discussions/20
* https://github.com/OAI/moonwalk/discussions/24
* https://github.com/OAI/moonwalk/discussions/31

As this change cannot be made in the 3.x line, I am closing it in favor of the Moonwalk discussions.
